### PR TITLE
ref(settings): Move various settings to new form components

### LINKED ISSUE #6315: After upgrading to 8.21, Sentry upgrade wizard hangs on auth.allow-registration
When upgrading my onpremise installation form 8.20 to 8.21 (using docker-compose), the wizard prompted me to set the “Anonymize Beacon” option. When clicking “Continue”, the following error message pops up:
> An invalid value for (auth.allow-registration) was passed to the server.

There was no such option `auth.allow-registration` set in either `sentry.conf.py` or `config.yml`.

The call to `https://sentry.example.com/api/0/internal/options/` returned a 400, with the following params:
```
system.admin-email: admin@example.com
system.url-prefix | https://sentry.example.com
auth.allow-registration | 0
beacon.anonymous | true
```
I worked around that by resending the request, changing 0 for False. My install is now fine

Running `sentry config get auth.allow-registration`, whether before the fix or after, returns:
```
        type: BOOLEAN
 from config: <not set>
     current: 0
```
This looks somehow similar to #5699.

Hope that helps!

### PR DESCRIPTION
**Pull request by dcramer**
Fixes #6315 and #5699

### CODE REVIEW

File: src/sentry/static/sentry/app/components/forms/formField.jsx [hunk 1]
```diff
@@ -31,22 +31,27 @@ export default class FormField extends React.Component {
...
+    // XXX(dcramer): defer the value until componentDidMount
+    // so we can propagate any changes upstream
+    this.state = {};
+  }
 
-    this.state = {
-      value: this.getValue(props, context)
-    };
+  componentDidMount() {
+    this.setValue(this.getValue(this.props, this.context));
```
Comment (dcramer) [inline src/sentry/static/sentry/app/components/forms/formField.jsx]: this is a little gross, but it solves the problem pretty well without introducing any additional complexity

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/static/sentry/app/components/forms/form.jsx [hunk 2]
```diff
@@ -83,11 +83,8 @@ export default class Form extends React.Component {
   };
 
   onFieldChange = (name, value) => {
-    this.setState({
-      data: {
-        ...this.state.data,
-        [name]: value
-      }
+    this.setState(state => {
```
Comment (dcramer) [inline src/sentry/static/sentry/app/components/forms/form.jsx]: this fixes some propagation bugs, but we could still do the copy inside of here safely if we REALLY wanted identity checks to change

File: src/sentry/static/sentry/app/components/forms/formField.jsx (hunk 1 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/components/forms/formField.jsx]: Yeah I find the data flow a bit confusing overall, this seems to be fine though

I also still think using OOP is very un-react-like, and will not lend itself well to future maintenance, especially for those that are more familiar w/ react patterns

**Review by billyvg — APPROVED** [review]

**Comment by dcramer** [comment]
This breaks the 'require changes' stuff in Form. I think I'll instead make an onInitialValue() hook for forms to call back to.

File: src/sentry/static/sentry/app/components/forms/form.jsx (hunk 2 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/components/forms/form.jsx]: This re-renders the entire form on every change

File: src/sentry/static/sentry/app/components/forms/formField.jsx (hunk 1 above)
Comment (dcramer) [reply src/sentry/static/sentry/app/components/forms/formField.jsx]: I think this is where OOP is the strongest, and javascript users should learn the pattern ;)

aside, uniflow feels like it'd be overly complex here, and all stores are globalized, so you'd effectively be just passing around callbacks (actions/dispatcher) anyways

File: src/sentry/static/sentry/app/components/forms/form.jsx (hunk 2 above)
Comment (dcramer) [reply src/sentry/static/sentry/app/components/forms/form.jsx]: yeah thats fine, downstream components could check their values

**Comment by dcramer** [comment]
I backtracked on coercion changes as it was complex, and instead opted to change BoolType to accept ints.

**Comment by dcramer** [comment]
need to write a test that confirms the value 'gAKJLg==' for 'sentry.allow-registration' works correctly

**Comment by dcramer** [comment]
I'm handing this over to @billyvg since my schedules a bit hectic and he can bring it home

**Pull request merged**